cpp/src/io/json/read_json.cu

```diff
@@ -209,10 +209,37 @@ datasource::owning_buffer<rmm::device_buffer> get_record_range_raw_input(
       reinterpret_cast<uint8_t*>(buffer.data()) + first_delim_pos + shift_for_nonzero_offset,
       next_delim_pos - first_delim_pos - shift_for_nonzero_offset);
   }
+
+  // Add delimiter to end of buffer iff
+  //        (i) We are reading till the end of the last source i.e. should_load_till_last_source is
+  //        true (ii) The last character in bufspan is not newline.
+  // For (ii) in the case of Spark, if the last character is not a newline, it could be the case
+  // that there are characters after the newline in the last record. We then consider those
+  // characters to be a part of a new (possibly empty) line.
+  size_t num_chars = readbufspan.size() - first_delim_pos - shift_for_nonzero_offset;
+  if (num_chars) {
+    char last_char;
+    CUDF_CUDA_TRY(cudaMemcpyAsync(&last_char,
+                                  reinterpret_cast<char*>(buffer.data()) + readbufspan.size() - 1,
+                                  sizeof(char),
+                                  cudaMemcpyDeviceToHost,
+                                  stream.value()));
+    stream.synchronize();
+    if (last_char != '\n') {
+      last_char = '\n';
```

Comment:
This is hardcoded to `\n`, should it be delimiter mentioned in json reader options instead?

Reply:
Qn: Does this force reader to do extra copy?

Reply:
Thank you, I've changed the hardcoded `\n` to the reader options delimiter. Also fixed it in a few other places.

Yes, we now have two extra copies between host and device each of size 1 byte. We also perform a stream sync between the copies. I'll run the JSON benchmarks to see what the impact of this change is.

Reply:
Once we merge #17028, we need to check the last character in the buffer only when both `nullify_empty_rows` and `recover_with_null` are enabled. Otherwise, I think we can always add a delimiter since empty rows are anyway ignored.